content/blog/2019-11-13-space-efficient-gc/index.md

```diff
@@ -0,0 +1,100 @@
+
+![003.png](./003.png)
+
+Several observations can be made based on the table:
+
+1. Blacklisting is effective in nearly eliminating all accidental retention caused by garbage collector conservativism.
+2. The numbers in the table are approximate as the results are not completely reproducible. This is due to the fact that the scanned part of the address space is polluted with UNIX environment variables. So they are specified as ranges.
+3. If all interior pointers are considered valid, it would be difficult to allocate individual objects larger than about 100 Kbytes without violating the blacklist constraint, or requesting memory from the OS at a garbage collector specified location.
+4. According to the paper, blacklist can be easily incorpoated into a garbage collecting allocator at *almost no performance cost*.
+
+### Other Sources of Excess Retention
+
+Another source of excess memory retention is due to the fact that conservative collectors usually have less information about variable liveness than conventional collectors. As it lacks these information, the allocator could come for help.
+
+Here are some useful techniques:
+- Have the allocator and collector carefully clean up after themselves, clearing local variables before function exit.
+- The allocator try to clear areas sometimes in the stack beyond the most recently activated frame.
+
+### Minor: Consequences of Misidentification
+
+Actually, the involved data structure can greatly influence the individual false reference. 
+
+For example, given a balanced binary tree, the expected number of vertices retained in a false reference is about the height of the tree, which is usually tolerable. Queses and lazy lists could exhisit much worse behavior as they grow without bound.
+
+A more common problem is the construction of largely stronly connected data structures, which could result in an unvounded memory leak if the structures are large enough. Take the data structures shown in the below graph as an example.
+
+![004.png](./004.png)
+
+They are a rectangular array of vertices, which are linked both horizontally and vertically. The structure can be accessed by traversing a row/column. The left shows an embedded link representation. A false reference in this structure would result in the retention of a large fraction of the whold structure. The right shows an embedded link representation of the same structure, with a separate link representation (represented by ovals in the fig). Thus, at most a single row/column is affected.
+
+Therefore, the embedded link version of some data structures would greatly help reduce storage retention and should be encouraged.
```

Comment:
FWIW, there is much more on this topic in [another paper by Hans](https://www.hpl.hp.com/techreports/2001/HPL-2001-251.pdf).